tt_metal/impl/device/device_impl.hpp

```diff
@@ -0,0 +1,309 @@
+    void end_trace(const uint8_t cq_id, const uint32_t tid);
+    void replay_trace(const uint8_t cq_id, const uint32_t tid, const bool blocking);
+    void release_trace(const uint32_t tid);
+    std::shared_ptr<TraceBuffer> get_trace(const uint32_t tid);
+
+    bool using_slow_dispatch() const;
+    void check_allocator_is_initialized() const;
+
+    // Checks that the given arch is on the given pci_slot and that it's responding
+    // Puts device into reset
+    bool initialize(const uint8_t num_hw_cqs, size_t l1_small_size, size_t trace_region_size, const std::vector<uint32_t> &l1_bank_remap = {}, bool minimal = false);
+    void initialize_cluster();
+    void initialize_allocator(size_t l1_small_size, size_t trace_region_size, const std::vector<uint32_t> &l1_bank_remap = {});
+    void initialize_build();
+    void build_firmware();
+    void initialize_firmware(CoreCoord phys_core, launch_msg_t *launch_msg);
+    void reset_cores();
+    void initialize_and_launch_firmware();
+    void init_command_queue_host();
+    void init_command_queue_device();
+    void initialize_synchronous_sw_cmd_queue();
+    void configure_kernel_variant(Program& program, string path, std::vector<uint32_t> compile_args, CoreCoord kernel_core, CoreCoord Kernel_physical_core,
+                                  CoreType dispatch_core_type, CoreCoord upstream_physical_core, CoreCoord downstream_physical_core, std::map<string, string> defines_in, NOC my_noc_index, NOC upstream_noc_index, NOC downstream_noc_index, bool is_active_eth_core = false);
+    void compile_command_queue_programs();
+    void configure_command_queue_programs();
+    void clear_l1_state();
+    void get_associated_dispatch_phys_cores(
+        std::unordered_map<chip_id_t, std::unordered_set<CoreCoord>> &my_dispatch_cores,
+        std::unordered_map<chip_id_t, std::unordered_set<CoreCoord>> &other_dispatch_cores);
+    std::pair<int, int> build_processor_type_to_index(JitBuildProcessorType t) const;
+
+    // Puts device into reset
+    bool close();
+    friend bool CloseDevice(Device *device);
+
+    // APIs to access this device's work executor
+    void push_work(std::function<void()>&& work, bool blocking = false);
+    void push_work(std::shared_ptr<std::function<void()>> work, bool blocking = false);
+    void synchronize();
+    void set_worker_mode(const WorkExecutorMode& mode);
+    void enable_async(bool enable);
+    WorkExecutorMode get_worker_mode() { return work_executor.get_worker_mode(); }
+    void set_worker_queue_mode(const WorkerQueueMode& mode) { this->work_executor.set_worker_queue_mode(mode); }
+    WorkerQueueMode get_worker_queue_mode() { return this->work_executor.get_worker_queue_mode(); }
+    // TODO: Uplift usage of friends. Buffer and Program just need access to allocator
+    friend class Buffer;
+    friend class Program;
+    friend class SystemMemoryManager;
+
+    static constexpr MemoryAllocator allocator_scheme_ = MemoryAllocator::L1_BANKING;
```

Comment:
can members go to private? or not yet? Not that opaque :D

Reply:
It's opaque in the sense that nothing outside of the TT Metal library sees its complete type. Changing the class definition in any way is out of scope for this PR.